## Preprocessing

In [20]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

# Second Optimization
Lets keep the NAME column to see it improve the results.

In [21]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df.drop(columns=['EIN'], axis=1, inplace= True)

In [22]:
# Determine the number of unique values in each column.
application_df.nunique()

NAME                      19568
APPLICATION_TYPE             17
AFFILIATION                   6
CLASSIFICATION               71
USE_CASE                      5
ORGANIZATION                  4
STATUS                        2
INCOME_AMT                    9
SPECIAL_CONSIDERATIONS        2
ASK_AMT                    8747
IS_SUCCESSFUL                 2
dtype: int64

In [23]:
#look at NAME value counts for binning
name_binning=application_df['NAME'].value_counts()
name_binning

PARENT BOOSTER USA INC                                                  1260
TOPS CLUB INC                                                            765
UNITED STATES BOWLING CONGRESS INC                                       700
WASHINGTON STATE UNIVERSITY                                              492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                          408
                                                                        ... 
ST LOUIS SLAM WOMENS FOOTBALL                                              1
AIESEC ALUMNI IBEROAMERICA CORP                                            1
WEALLBLEEDRED ORG INC                                                      1
AMERICAN SOCIETY FOR STANDARDS IN MEDIUMSHIP & PSYCHICAL INVESTIGATI       1
WATERHOUSE CHARITABLE TR                                                   1
Name: NAME, Length: 19568, dtype: int64

In [24]:
# Choose a cutoff value and create a list of Name to be replaced
# use the variable name `name_to_replace`
name_to_replace= list(name_binning[name_binning < 100].index)

# Replace in dataframe
for name in name_to_replace:
    application_df['NAME'] = application_df['NAME'].replace(name,"Other")

# Check to make sure binning was successful
application_df['NAME'].value_counts()

Other                                                                 25987
PARENT BOOSTER USA INC                                                 1260
TOPS CLUB INC                                                           765
UNITED STATES BOWLING CONGRESS INC                                      700
WASHINGTON STATE UNIVERSITY                                             492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                         408
PTA TEXAS CONGRESS                                                      368
SOROPTIMIST INTERNATIONAL OF THE AMERICAS INC                           331
ALPHA PHI SIGMA                                                         313
TOASTMASTERS INTERNATIONAL                                              293
MOST WORSHIPFUL STRINGER FREE AND ACCEPTED MASONS                       287
LITTLE LEAGUE BASEBALL INC                                              277
INTERNATIONAL ASSOCIATION OF LIONS CLUBS                                266
MOMS CLUB   

In [7]:
# Look at APPLICATION_TYPE value counts for binning
app_type_binning= application_df['APPLICATION_TYPE'].value_counts()
app_type_binning

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [8]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace= list(app_type_binning[app_type_binning < 500].index)

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [9]:
# Look at CLASSIFICATION value counts for binning
classification_binning= application_df['CLASSIFICATION'].value_counts()
classification_binning

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [10]:
# You may find it helpful to look at CLASSIFICATION value counts >1
class_binning_grt1= classification_binning[classification_binning > 1]
class_binning_grt1

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [11]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace = list(classification_binning[classification_binning < 150].index)

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
Other     1003
C7000      777
C1700      287
C4000      194
Name: CLASSIFICATION, dtype: int64

In [25]:
# Convert categorical data to numeric with `pd.get_dummies`
df_dummies= pd.get_dummies(application_df, columns=['NAME','APPLICATION_TYPE', 'AFFILIATION', 'CLASSIFICATION', 'USE_CASE', 'ORGANIZATION', 'INCOME_AMT', 'SPECIAL_CONSIDERATIONS'])
df_dummies.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  NAME_ALPHA PHI SIGMA  \
0       1     5000              1                     0   
1       1   108590              1                     0   
2       1     5000              0                     0   
3       1     6692              1                     0   
4       1   142590              1                     0   

   NAME_AMATEUR ATHLETIC UNION OF THE UNITED STATES INC  \
0                                                  0      
1                                                  0      
2                                                  0      
3                                                  0      
4                                                  0      

   NAME_AMERICAN ASSOCIATION OF UNIVERSITY WOMEN  NAME_CIVITAN INTERNATIONAL  \
0                                              0                           0   
1                                              0                           0   
2                                              0                           0   
3                                              0                           0   
4                                              0                           0   

   NAME_DEMOLAY INTERNATIONAL  \
0                           0   
1                           0   
2                           0   
3                           0   
4                           0   

   NAME_FARMERS EDUCATIONAL AND COOPERATIVE UNION OF AMERICA  \
0                                                  0           
1                                                  0           
2                                                  0           
3                                                  0           
4                                                  0           

   NAME_HABITAT FOR HUMANITY INTERNATIONAL INC  ...  INCOME_AMT_1-9999  \
0                                            0  ...                  0   
1                                            0  ...                  1   
2                                            0  ...                  0   
3                                            0  ...                  0   
4                                            0  ...                  0   

   INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
0                       0                         0                   0   
1                       0                         0                   0   
2                       0                         0                   0   
3                       1                         0                   0   
4                       0                         1                   0   

   INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  INCOME_AMT_50M+  \
0                 0                       0                0   
1                 0                       0                0   
2                 0                       0                0   
3                 0                       0                0   
4                 0                       0                0   

   INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  
0                  0                         1                         0  
1                  0                         1                         0  
2                  0                         1                         0  
3                  0                         1                         0  
4                  0                         1                         0  

[5 rows x 148 columns]

In [26]:
# Split our preprocessed data into our features and target arrays
X= df_dummies.drop('IS_SUCCESSFUL', axis=1).values
y= df_dummies['IS_SUCCESSFUL'].values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test= train_test_split(X, y, random_state=42, stratify=y)

In [27]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [28]:
X_train.shape

(25724, 147)

## Compile, Train and Evaluate the Model

# Second Optimization

In [29]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
num_input= X_train.shape[1]
first_layer_units= 100

nn3 = tf.keras.models.Sequential()

# First hidden layer
nn3.add(tf.keras.layers.Dense(input_dim=num_input, units=first_layer_units, activation='relu'))

# Second hidden layer
nn3.add(tf.keras.layers.Dense(units=50, activation='relu'))

# Third hidden layer
nn3.add(tf.keras.layers.Dense(units=30, activation='relu'))

# Output layer
nn3.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn3.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 100)               14800     
                                                                 
 dense_1 (Dense)             (None, 50)                5050      
                                                                 
 dense_2 (Dense)             (None, 30)                1530      
                                                                 
 dense_3 (Dense)             (None, 1)                 31        
                                                                 
Total params: 21411 (83.64 KB)
Trainable params: 21411 (83.64 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [30]:
# Compile the model
nn3.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [31]:
# Train the model
nn3.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
804/804 [==============================] - 4s 3ms/step - loss: 0.5132 - accuracy: 0.7449
Epoch 2/50
804/804 [==============================] - 2s 3ms/step - loss: 0.4909 - accuracy: 0.7577
Epoch 3/50
804/804 [==============================] - 2s 3ms/step - loss: 0.4877 - accuracy: 0.7593
Epoch 4/50
804/804 [==============================] - 2s 3ms/step - loss: 0.4843 - accuracy: 0.7612
Epoch 5/50
804/804 [==============================] - 2s 3ms/step - loss: 0.4837 - accuracy: 0.7616
Epoch 6/50
804/804 [==============================] - 2s 3ms/step - loss: 0.4825 - accuracy: 0.7636
Epoch 7/50
804/804 [==============================] - 2s 3ms/step - loss: 0.4812 - accuracy: 0.7626
Epoch 8/50
804/804 [==============================] - 2s 3ms/step - loss: 0.4806 - accuracy: 0.7635
Epoch 9/50
804/804 [==============================] - 2s 3ms/step - loss: 0.4795 - accuracy: 0.7644
Epoch 10/50
804/804 [==============================] - 2s 3ms/step - loss: 0.4785 - accuracy: 0.7658

In [32]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn3.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5042 - accuracy: 0.7517 - 765ms/epoch - 3ms/step
Loss: 0.5042019486427307, Accuracy: 0.751720130443573


# The second optimization model results in 0.7517 Accuracy !
    - Included the NAME values for binning.
    - Increased the hidden layer.
    - Decreased the Epoch to 50.

In [34]:
# Export our model to HDF5 file
nn3.save('models/AlphabetSoupClarity_Optimization.h5')

c:\Users\aksha\anaconda3\lib\site-packages\keras\src\engine\training.py:3000: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


# First Optimization 

In [17]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
num_input= X_train.shape[1]
first_layer_units= 100

nn2 = tf.keras.models.Sequential()

# First hidden layer
nn2.add(tf.keras.layers.Dense(input_dim=num_input, units=first_layer_units, activation='relu'))

# Second hidden layer
nn2.add(tf.keras.layers.Dense(units=50, activation='relu'))

# Third hidden layer
nn2.add(tf.keras.layers.Dense(units=30, activation='relu'))

# Output layer
nn2.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn2.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 100)               4500      
                                                                 
 dense_1 (Dense)             (None, 50)                5050      
                                                                 
 dense_2 (Dense)             (None, 30)                1530      
                                                                 
 dense_3 (Dense)             (None, 1)                 31        
                                                                 
Total params: 11111 (43.40 KB)
Trainable params: 11111 (43.40 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [18]:
# Compile the model
nn2.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [21]:
# Train the model
nn2.fit(X_train_scaled, y_train, epochs=150)

Epoch 1/150
804/804 [==============================] - 2s 3ms/step - loss: 0.5310 - accuracy: 0.7420
Epoch 2/150
804/804 [==============================] - 2s 3ms/step - loss: 0.5306 - accuracy: 0.7423
Epoch 3/150
804/804 [==============================] - 2s 3ms/step - loss: 0.5306 - accuracy: 0.7432
Epoch 4/150
804/804 [==============================] - 2s 3ms/step - loss: 0.5306 - accuracy: 0.7421
Epoch 5/150
804/804 [==============================] - 2s 3ms/step - loss: 0.5308 - accuracy: 0.7426
Epoch 6/150
804/804 [==============================] - 2s 2ms/step - loss: 0.5309 - accuracy: 0.7426
Epoch 7/150
804/804 [==============================] - 2s 3ms/step - loss: 0.5307 - accuracy: 0.7429
Epoch 8/150
804/804 [==============================] - 2s 3ms/step - loss: 0.5308 - accuracy: 0.7422
Epoch 9/150
804/804 [==============================] - 2s 3ms/step - loss: 0.5310 - accuracy: 0.7422
Epoch 10/150
804/804 [==============================] - 2s 2ms/step - loss: 0.5306 - accura

In [22]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn2.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5965 - accuracy: 0.7257 - 547ms/epoch - 2ms/step
Loss: 0.5964804291725159, Accuracy: 0.7257142663002014


In [38]:
# First Optimization Model did only 0.7257 Accuracy .
    - Increased the number of hidden layers.
    - Decreased the units.
    - Increased the Epoch to 150.